model/jpa/src/main/java/org/keycloak/models/jpa/ClientAdapter.java

```diff
@@ -613,27 +598,45 @@ public void setServiceAccountsEnabled(boolean serviceAccountsEnabled) {
 
     @Override
     public RoleModel getRole(String name) {
-        return session.realms().getClientRole(realm, this, name);
+        for(RoleModel role : getRoles()) {
+            if(name != null && name.equals(role.getName())) {
+                return role;
+            }
+        }
+        return null;
     }
 
     @Override
     public RoleModel addRole(String name) {
-        return session.realms().addClientRole(realm, this, name);
+        return addRole(KeycloakModelUtils.generateId(), name);
```

Comment:
Not 100% sure, but I think this can result in missing cache invalidations. The reason why we use methods on the provider-level is to ensure that cache invalidations are not missed. Note that it can be the case when for example, client is not cached, but client roles are cached. In this case, roles list need to be invalidated from the cache. Etc.

Reply:
You are right. I reverted my changes on client roles.